### PR TITLE
Pull Request for adding Token and API key authentication mechanism

### PR DESCRIPTION
**Pull request by arpitamishra-macrometa**
This PR contains below implementation -
1. JWT Token based authentication 
2. API Key based authetication

### CODE REVIEW

File: src/main/java/com/c8db/C8DB.java [hunk 1]
```diff
@@ -602,12 +614,11 @@ public synchronized C8DB build() {
             final ConnectionFactory connectionFactory = (protocol == null || Protocol.VST == protocol)
                     ? new VstConnectionFactorySync(host, timeout, connectionTtl, useSsl, sslContext)
                     : new HttpConnectionFactory(timeout, user, password, email, jwtAuth, useSsl, sslContext, custom, protocol,
-                            connectionTtl, httpCookieSpec);
+                            connectionTtl, httpCookieSpec,jwtToken,apiKey);
```
Comment (grainier-macrometa) [inline src/main/java/com/c8db/C8DB.java:617]: @arpitamishra-macrometa - Small formatting issue here.

File: src/main/java/com/c8db/internal/InternalC8DBBuilder.java [hunk 2]
```diff
@@ -157,6 +161,14 @@ protected void setEmail(String email) {
         this.email = email;
     }
 
+    protected void setJwtToken(String jwtToken) {
```
Comment (grainier-macrometa) [inline src/main/java/com/c8db/internal/InternalC8DBBuilder.java:164]: @arpitamishra-macrometa - Add modifications copyright (https://wiki.macrometa.io/eng-docs/ApacheLicensev20Software_modifications)

File: src/main/java/com/c8db/internal/http/HttpConnection.java [hunk 3]
```diff
@@ -86,25 +86,33 @@
...
+    private static volatile String jwt;
+    private static volatile String apiKey;
+
 
     private HttpConnection(final HostDescription host, final Integer timeout, final String user, final String password,
                            final String email, final Boolean jwtAuthEnabled, final Boolean useSsl,
                            final SSLContext sslContext, final C8Serialization util,
-                           final Protocol contentType, final Long ttl, final String httpCookieSpec) {
+                           final Protocol contentType, final Long ttl, final String httpCookieSpec,
+                           String jwt,String apiKey,final Boolean apiKeyEnabled) {
```
Comment (grainier-macrometa) [inline src/main/java/com/c8db/internal/http/HttpConnection.java:102]: @arpitamishra-macrometa - Formatting issues (no space after `,`).

File: src/main/java/com/c8db/internal/http/HttpConnection.java [hunk 4]
```diff
@@ -212,17 +220,23 @@ public Response execute(final Request request) throws C8DBException, IOException
...
         addHeader(request, httpRequest);
-        if (jwtAuthEnabled) {
+        if (jwtAuthEnabled && !apiKeyEnabled) {
             if (jwt == null) {
                 addJWT();
             }
+            LOGGER.info("Using JWT for authentication.");
             httpRequest.addHeader("Authorization", "bearer " + jwt);
+        }else if(apiKeyEnabled && apiKey != null){
+            LOGGER.info("Using API Key for authenication.");
```
Comment (grainier-macrometa) [inline src/main/java/com/c8db/internal/http/HttpConnection.java:235]: @arpitamishra-macrometa - Just to clarify, if both `jwt` and `apiKey` enabled, you opt into `apiKey`, right?

File: src/main/java/com/c8db/internal/http/HttpConnection.java [hunk 5]
```diff
@@ -86,25 +86,33 @@
...
     private final String email;
     private final Boolean jwtAuthEnabled;
+    private Boolean apiKeyEnabled;
     private final C8Serialization util;
     private final Boolean useSsl;
     private final Protocol contentType;
     private final HostDescription host;
-    private volatile String jwt;
+    private static volatile String jwt;
+    private static volatile String apiKey;
```
Comment (grainier-macrometa) [inline src/main/java/com/c8db/internal/http/HttpConnection.java:95]: @arpitamishra-macrometa - Why these are `static`? If a user creates 2 c8db instances for two DBs, won't this be an issue?

File: src/main/java/com/c8db/internal/http/HttpConnection.java [hunk 6]
```diff
@@ -86,25 +86,33 @@
     private final String password;
     private final String email;
     private final Boolean jwtAuthEnabled;
+    private Boolean apiKeyEnabled;
```
Comment (StoyanTc) [inline src/main/java/com/c8db/internal/http/HttpConnection.java:89]: @arpitamishra-macrometa - Please get rid of this variable, it's redundant.

File: src/main/java/com/c8db/internal/http/HttpConnection.java (hunk 4 above)
Comment (StoyanTc) [reply src/main/java/com/c8db/internal/http/HttpConnection.java:235]: @grainier-macrometa - Keeping the implementation close to the original, it would better to opt for `jwt`.

**Review by StoyanTc — COMMENTED** [review]
@arpitamishra-macrometa - Let's make the implementation a bit simpler and closer to the original implementation, assuming having less changes less problems. That's why I'd like to propose you to get rid of `apiKeyEnabled`, it's also redundant in my opinion.

File: src/main/java/com/c8db/internal/http/HttpConnection.java (hunk 4 above)
Comment (arpitamishra-macrometa) [reply src/main/java/com/c8db/internal/http/HttpConnection.java:235]: Incorporated the implementation - if both jwt and apiKey are provided, then jwt takes preference.

File: src/main/java/com/c8db/internal/http/HttpConnection.java (hunk 5 above)
Comment (arpitamishra-macrometa) [reply src/main/java/com/c8db/internal/http/HttpConnection.java]: Removed static and replaced with instance variable.

File: src/main/java/com/c8db/internal/http/HttpConnection.java (hunk 6 above)
Comment (arpitamishra-macrometa) [reply src/main/java/com/c8db/internal/http/HttpConnection.java]: Done.

File: src/main/java/com/c8db/internal/http/HttpConnection.java [hunk 7]
```diff
@@ -212,17 +220,23 @@ public Response execute(final Request request) throws C8DBException, IOException
...
         }
         addHeader(request, httpRequest);
-        if (jwtAuthEnabled) {
+        if (jwtAuthEnabled && !apiKeyEnabled) {
             if (jwt == null) {
                 addJWT();
             }
+            LOGGER.info("Using JWT for authentication.");
             httpRequest.addHeader("Authorization", "bearer " + jwt);
+        }else if(apiKeyEnabled && apiKey != null){
```
[the remark this answers is not in the document]
Comment (arpitamishra-macrometa) [reply src/main/java/com/c8db/internal/http/HttpConnection.java]: Done

File: src/main/java/com/c8db/internal/InternalC8DBBuilder.java [hunk 8]
```diff
@@ -138,7 +141,9 @@ protected void loadProperties(final Properties properties) {
         user = loadUser(properties, user);
         password = loadPassword(properties, password);
         email = loadEmail(properties, email);
+        jwtToken = loadJWTToken(properties,jwtToken);
```
Comment (grainier-macrometa) [inline src/main/java/com/c8db/internal/InternalC8DBBuilder.java:144]: @arpitamishra-macrometa - Few formatting issues (no whitespace after `,`)

File: src/main/java/com/c8db/internal/InternalC8DBBuilder.java [hunk 9]
```diff
@@ -205,6 +218,32 @@ protected void serializer(final C8Serializer serializer) {
         this.serializer = serializer;
     }
 
+    protected HostHandler createHostHandler(final HostResolver hostResolver) {
```
Comment (grainier-macrometa) [inline src/main/java/com/c8db/internal/InternalC8DBBuilder.java:221]: @arpitamishra-macrometa - If possible, shall we revert this change `createHostHandler`?

**Review by grainier-macrometa — APPROVED** [review]
Added 2 comments. Other than that, LGTM.

File: src/main/java/com/c8db/internal/http/HttpConnection.java [hunk 10]
```diff
@@ -212,17 +219,26 @@ public Response execute(final Request request) throws C8DBException, IOException
...
         final HttpRequestBase httpRequest = buildHttpRequestBase(request, url);
         httpRequest.setHeader("User-Agent", "Mozilla/5.0 (compatible; C8DB-JavaDriver/1.1; +http://mt.orz.at/)");
+
         if (contentType == Protocol.HTTP_VPACK) {
             httpRequest.setHeader("Accept", "application/x-velocypack");
         }
         addHeader(request, httpRequest);
         if (jwtAuthEnabled) {
-            if (jwt == null) {
+            if (apiKey != null && jwt == null) {  //Use API key onlu if API Key is provided
```
Comment (StoyanTc) [inline src/main/java/com/c8db/internal/http/HttpConnection.java:228]: @arpitamishra-macrometa - Please use `StringUtils.isNotEmpty(apiKey)` to cover both cases - `null` and `""`?
There is a typo in your comment out.

File: src/main/java/com/c8db/internal/http/HttpConnectionFactory.java [hunk 11]
```diff
@@ -33,12 +33,12 @@
...
             final Boolean jwtAuth, final Boolean useSsl, final SSLContext sslContext, final C8Serialization util,
-            final Protocol protocol, final Long connectionTtl, String httpCookieSpec) {
+            final Protocol protocol, final Long connectionTtl, String httpCookieSpec, final String jwtToken, final String apiKey) {
         super();
         builder = new HttpConnection.Builder().timeout(timeout).user(user).password(password).email(email)
                 .jwtAuthEnabled(jwtAuth).useSsl(useSsl).sslContext(sslContext).serializationUtil(util)
-                .contentType(protocol).ttl(connectionTtl).httpCookieSpec(httpCookieSpec);
-
+                .contentType(protocol).ttl(connectionTtl).httpCookieSpec(httpCookieSpec).jwt(jwtToken)
+                .apiKey(apiKey);
```
Comment (StoyanTc) [inline src/main/java/com/c8db/internal/http/HttpConnectionFactory.java:41]: @arpitamishra-macrometa - Please keep that call on the line above - you have already such long lines.

**Review by StoyanTc — COMMENTED** [review]
@arpitamishra-macrometa - Just few minor fixes and you'll be ready.

File: src/main/java/com/c8db/internal/InternalC8DBBuilder.java (hunk 9 above)
Comment (arpitamishra-macrometa) [reply src/main/java/com/c8db/internal/InternalC8DBBuilder.java:221]: This was automatically done because I added my implementation right after jwt

File: src/main/java/com/c8db/internal/http/HttpConnection.java (hunk 10 above)
Comment (arpitamishra-macrometa) [reply src/main/java/com/c8db/internal/http/HttpConnection.java:228]: Added the same for apiKey

File: src/main/java/com/c8db/internal/http/HttpConnectionFactory.java (hunk 11 above)
Comment (arpitamishra-macrometa) [reply src/main/java/com/c8db/internal/http/HttpConnectionFactory.java:41]: Done. I just kept it that way to make the code look cleaner.

File: src/main/java/com/c8db/internal/InternalC8DBBuilder.java (hunk 8 above)
Comment (arpitamishra-macrometa) [reply src/main/java/com/c8db/internal/InternalC8DBBuilder.java:144]: Done

**Review by StoyanTc — APPROVED** [review]

File: src/main/java/com/c8db/internal/InternalC8DBBuilder.java (hunk 9 above)
Comment (StoyanTc) [reply src/main/java/com/c8db/internal/InternalC8DBBuilder.java:221]: @grainier-macrometa - I cannot understand, what's the issue here?

**Review by StoyanTc — APPROVED** [review]

File: src/main/java/com/c8db/internal/InternalC8DBBuilder.java (hunk 9 above)
Comment (grainier-macrometa) [reply src/main/java/com/c8db/internal/InternalC8DBBuilder.java:221]: @StoyanTc, @arpitamishra-macrometa  - It's not an issue. However, git history will re-write that change as a part of `Introducing Token and API key authentication mechanism`. So, If we annotate and refer to it later, annotation might not make sense.

**Review by grainier-macrometa — APPROVED** [review]

File: src/main/java/com/c8db/internal/InternalC8DBBuilder.java (hunk 9 above)
Comment (StoyanTc) [reply src/main/java/com/c8db/internal/InternalC8DBBuilder.java:221]: @grainier-macrometa - I do agree.
